fix: environment for executing code

nautics889 opened the pull request:
This PR should resolve the problem with environment described in issue #400.

For now i've just added `matplotlib.pyplot` as "plt" in `_get_environment()` method. The package is present in dependency list, code from the issue runs without failures, it must works OK.

However it's still unclear for me, should we add the rest of popular packages, such as `numpy` as "np" and so on? Although `PandasAI` already has [`import_dependency()`](https://github.com/gventuri/pandas-ai/blob/main/pandasai/helpers/_optional.py#L45) util-function, and this logic in theory must have to avoid this problem, this  couldn't work because of (regarding to [the issue](https://github.com/gventuri/pandas-ai/issues/400)) HuggingFace just had returned code using `plt` variable without import statement — just something like `plt.show()`, and this caused `NameError`, since there was no such thing as `plt` in the context.

I've some idea for this cases, when API provider "forgets" to add import-statements. Like, for example, force adding `import <var_name>`  to the code when executing, if `NameError: name <var_name> is not defined` occured for example (btw this wouldn't work with aliases, but anyway can help with some other cases). But i'd like to hear opinion from others first.
@gventuri, @mspronesti welcome to discuss=)

___

* (fix): set `matplot.pyplot` package under "plt" alias in `_get_environment()`

___

- [x] closes #400
- [x] closes #430 
- [x] Tests added and passed if fixing a bug or adding a new feature
- [x] All [code checks passed](https://github.com/gventuri/pandas-ai#contributing).

nautics889 commented:
Thank for replying, @gventuri 
___

>it's more a game of figuring out how to import a given library if it hasn't been imported by the LLM for some reason

Yes, absolutely agree, that would be great if we can, for example, try to import a package if there were raised `NameError`. I'll upload some updates within this PR if you don't mind.
___
However there is still another point of the problem: how should we handle cases when LLM gives a code snippet **using a common alias**, like `np` for `numpy` or `plt` for `matplotlib.plt`, and without import statements (`import numpy as np` and so on). This is exactly the case appeared in https://github.com/gventuri/pandas-ai/issues/400.

nautics889 commented:
recheck

nautics889 commented:
Huh, as it turns out, `NameError` for python 3.9 does not have attribute `.name`, that's why tests failing only for 3.9. Already working on it...

nautics889 commented:
@gventuri could you pls take a look?

___

Here's a brief description of what i've made within this PR:

* Split up code direct code execution and handling of errors to different methods. Drag them out from `run_code()`.
* Method `_execute_catching_errors()` just calls `exec(code, env)` wrapped by try-except, then returns exception if any errors appeared. Otherwise returns None.
* Method `handle_error()` is being called when at the very first execution of a code some exception was returned. It check if the error was an instance of `NameError`. If so, tries to fix that; otherwise continues troubleshooting. If `use_error_correction_framework` is True, tries to re-execute code, given by `_retry_run_code()`. Sure, it tries to do this `self._max_tries` times, or until `_execute_catching_errors()` returns None (meaning there was no exception during execution).

I've been trying to implement this small refactoring seamlessly, so essentially, there is no any change of original logic, except that a handling of `NameError` was added in addition to "error correction framework".

p. s. sorry for massive git diff

gventuri commented on pandasai/__init__.py:
```diff
@@ -642,6 +643,7 @@ def _get_environment(self) -> dict:
 
         return {
             "pd": pd,
+            "plt": matplotlib.pyplot,
```
We don't need this, as we automatically import the relevant dependencies (with the right alias) with `import_dependency` a few lines below.

nautics889 replied:
> We don't need this, as we automatically import the relevant dependencies (with the right alias) with `import_dependency` a few lines below.

Roger that, with a next commit.

gventuri replied:
Great, thanks a lot!

gventuri commented on pandasai/__init__.py:
```diff
@@ -661,7 +663,7 @@ def _retry_run_code(self, code: str, e: Exception, multiple: bool = False):
             e (Exception): An exception
             multiple (bool): A boolean to indicate if the code is for multiple
-            dataframes
+                             dataframes
```
This is probably a leftover?

nautics889 replied:
> This is probably a leftover?

Not really... This comes mostly from a matter of taste whether to make visual indent for multilines description of arguments/possible exceptions/notes/return values. 
___
Let me show...

```python
def foobar(arg_one, arg_two):
    """
    ...

    Args:
        arg_one (str): Lorem ipsum dolor sit amet, consectetur adipiscing elit. 
        Vivamus bibendum mollis odio. Ut tellus nunc, tempor a elit sit amet, 
        fermentum eleifend justo. Nulla sit amet diam in erat finibus 
        arg_two (Exception): Lorem ipsum dolor sit amet, consectetur adipiscing elit. 
        Vivamus bibendum mollis odio. Ut tellus nunc, tempor a elit sit amet

    ...
    """
```
(no indend after new line for args verbal descriptions, kind of difficult to read as for me :) )

```python
def foobar(arg_one, arg_two):
    """
    ...

    Args:
        arg_one (str): Lorem ipsum dolor sit amet, consectetur adipiscing elit.
            Vivamus bibendum mollis odio. Ut tellus nunc, tempor a elit sit amet,
            fermentum eleifend justo. Nulla sit amet diam in erat finibus
        arg_two (Exception): Lorem ipsum dolor sit amet, consectetur adipiscing elit.
            Vivamus bibendum mollis odio. Ut tellus nunc, tempor a elit sit amet

    ...
    """
```
(indent is present for each new line in description, this should make docstring more readable as it separates arguments from each other)

___

References: [example of Google-style docstrings for python](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)
![image](https://github.com/gventuri/pandas-ai/assets/31508183/2132a20d-72dc-4cb6-9c26-c337278c6999)
___

HOWEVER, now i think it'd be better to make changes like that in a separate PR created specifically for docs/doc-style improvement. Of course, if you don't mind it.

gventuri replied:
Got it, the few lines shown without expanding weren't providing me enough context! Makes a lot of sense then, thanks for clarifying!

gventuri commented on tests/test_pandasai.py:
```diff
@@ -146,6 +146,43 @@ def test_run_without_verbose(self, pandasai, llm):
+        assert code == pandasai.handle_error(
+            exc, code, environment, use_error_correction_framework=False, multiple=False
+        )
+        assert getattr(environment.get("os"), "__name__", None) == "os"
```
Os is actually one of the libraries that we never want to allow, since it could potentially expose env variables and raise many more concerns 😅

nautics889 replied:
> Os is actually one of the libraries that we never want to allow, since it could potentially expose env variables and raise many more concerns 😅

Great point, thanks! Will try to find something else from stdlib...

gventuri replied:
Great idea (and that makes me think that we might even need a "blacklist" list to prevent such dangerous libs from being exposed).

gventuri commented on pandasai/__init__.py:
```diff
@@ -703,6 +705,115 @@ def _retry_run_code(self, code: str, e: Exception, multiple: bool = False):
+
+            return exc
+
+    def handle_error(
```
IMO here we should definitely check whether what we are trying to import is indeed a whitelisted library, or it's just too dangerous!

nautics889 replied:
> we should definitely check whether what we are trying to import is indeed a whitelisted library
EXACTLY, had the same thought, i just needed to hear confirmation for this from you

gventuri replied:
Sure, let's go for the allowing only whitelisted then! Thanks a lot!

gventuri commented:
@nautics889 thanks a lot for the MR. Just left a couple of comments, mostly about security!

nautics889 commented:
@gventuri thank you for the review!

I'm already working on a new commit; i've left some replies, will appreciate if you find time to take a look.

nautics889 commented:
@gventuri 
I've uploaded an update. Added all the changes you've mentioned. 
____
However, there is an important point i've discovered during i was making the change. 

According to the logic of `run_code()` which we have now, when pandasai caught an exception during code running (and `use_error_correction_framework` is `True` of course), it tries to run code over and over again with the error correction framework, until in exceeds `_max_retries`. And in case when **the error correction framework can't manage to fix that** within `_max_retries` numbers of efforts, **non of exception will be raised**, silently, just the last form of code will be returned and processed further (in practice, sure, it definitely causes further errors). Is that logic correct and it's supposed to be like this?

So, i decided to raise the latest exception for such cases (sources [here](https://github.com/gventuri/pandas-ai/pull/419/commits/030b3aed5944f3b078f314a744674a7e7e0f09b6#diff-62b472d83558337a7726ff6758a8c431481784a74add715c96e2fed8bedbbd63R812)):
```python
        for retry_num in range(self._max_retries):
            code = self._retry_run_code(code, exc, multiple)
            caught_error = self._execute_catching_errors(code, environment)
            if caught_error is None:
                break
            self.log(
                f"Failed to execute code with a correction framework "
                f"[retry number: {retry_num + 1}]",
                level=logging.WARNING,
            )
        else:
            # `else` will only starts when `break` above didn't run in all the loop, i. e. ECF couldn't fix the error
            self.log(
                f"Unable to fix {exc.__class__.__name__} with error "
                f"correction framework",
                level=logging.ERROR,
            )
            raise exc

```
Also i added [tests for such cases](https://github.com/gventuri/pandas-ai/pull/419/commits/030b3aed5944f3b078f314a744674a7e7e0f09b6#diff-62b472d83558337a7726ff6758a8c431481784a74add715c96e2fed8bedbbd63R812).

___

Well, there is something else interesting i've encountered on. When i added the changes which i've described above, i found several tests cases failing. For example [this one](https://github.com/gventuri/pandas-ai/pull/419/commits/030b3aed5944f3b078f314a744674a7e7e0f09b6#diff-92144ec0277f612e0b022a692801edf93584cc8ecb551e0050411ba44207b3f5L389).
The reason for they were failing:
It's unable to create numpy's `array` with not passing any iterable object to it's `__init__()`.
_incorrect way:_
```python
In [1]: import numpy as np
In [2]: np.array()
Traceback (most recent call last):
  File "***\venvs\pandas-ai-venv\Lib\site-packages\IPython\core\interactiveshell.py", line 3508, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-5-e4f3b47dc252>", line 1, in <module>
    np.array()
TypeError: array() missing required argument 'object' (pos 0)
```
_correct way:_
```python
In [1]: np.array([1, 2, 3])
Out[2]: array([1, 2, 3])
```
The reason why we didn't see those tests failed before:
As i've said
>**non of exception will be raised**, silently, just the last form of code will be returned and processed further
Considering this, tests were keeping silence, even when some exception actually had been raised when `exec()` code from LLM mocks.

___

Let me know if i'm wrong and exception suppressing should be returned back.

P. S. Apologize again for miscellaneous issues within a single PR and large git diff.

gventuri commented:
@nautics889 impressive job, thanks a lot for also fixing all the other edge cases you've highlighted! Having access to the last exception is a game-changer for debug! Merging!
